dhall-json/json-to-dhall/Main.hs

```diff
@@ -0,0 +1,586 @@
+
+> $ json-to-dhall Bool <<< 'true'
+> True
+> $ json-to-dhall Bool <<< 'false'
+> False
+
+    JSON numbers translate to Dhall numbers:
+
+> $ json-to-dhall Integer <<< 2
+> +2
+> $ json-to-dhall Natural <<< 2
+> 2
+> $ json-to-dhall Double <<< -2.345
+> -2.345
+
+    Dhall @Text@ corresponds to JSON text:
+
+> $ json-to-dhall Text <<< '"foo bar"'
+> "foo bar"
+
+
+== Lists and records
+
+    Dhall @List@s correspond to JSON lists:
+
+> $ json-to-dhall 'List Integer' <<< '[1, 2, 3]'
+> [ +1, +2, +3 ]
+
+
+    Dhall __records__ correspond to JSON records:
+
+> $ json-to-dhall '{foo : List Integer}' <<< '{"foo": [1, 2, 3]}'
+> { foo = [ +1, +2, +3 ] }
+
+
+    Note, that by default, only the fields required by the Dhall type argument are parsed (as you commonly will not need all the data), the remaining ones being ignored:
+
+> $ json-to-dhall '{foo : List Integer}' <<< '{"foo": [1, 2, 3], "bar" : "asdf"}'
+> { foo = [ +1, +2, +3 ] }
+
+
+    If you do need to make sure that Dhall fully reflects JSON record data comprehensively, @--records-strict@ flag should be used:
+
+> $ json-to-dhall --records-strict '{foo : List Integer}' <<< '{"foo": [1, 2, 3], "bar" : "asdf"}'
+> Error: Key(s) @bar@ present in the JSON object but not in the corresponding Dhall record. This is not allowed in presence of --records-strict:
+
+
+    By default, JSON key-value arrays will be converted to Dhall records:
+
+> $ json-to-dhall '{ a : Integer, b : Text }' <<< '[{"key":"a", "value":1}, {"key":"b", "value":"asdf"}]'
+> { a = +1, b = "asdf" }
+
+
+    Attempting to do the same with @--no-keyval-arrays@ on will result in error:
+
+> $ json-to-dhall --no-keyval-arrays '{ a : Integer, b : Text }' <<< '[{"key":"a", "value":1}, {"key":"b", "value":"asdf"}]'
+> Error: JSON (key-value) arrays cannot be converted to Dhall records under --no-keyval-arrays flag:
+
+    Conversion of the homogeneous JSON maps to the corresponding Dhall association lists by default:
+
+> $ json-to-dhall 'List { mapKey : Text, mapValue : Text }' <<< '{"foo": "bar"}'
+> [ { mapKey = "foo", mapValue = "bar" } ]
+
+    Flag @--no-keyval-maps@ switches off this mechanism (if one would ever need it):
+
+> $ json-to-dhall --no-keyval-maps 'List { mapKey : Text, mapValue : Text }' <<< '{"foo": "bar"}'
+> Error: Homogeneous JSON map objects cannot be converted to Dhall association lists under --no-keyval-arrays flag
+
+
+== Optional values and unions
+
+    Dhall @Optional@ Dhall type allows null or missing JSON values:
+
+> $ json-to-dhall "Optional Integer" <<< '1'
+> Some +1
+
+> $ json-to-dhall "Optional Integer" <<< null
+> None
+
+> $ json-to-dhall '{ a : Integer, b : Optional Text }' <<< '{ "a": 1 }'
+{ a = +1, b = None Text }
+
+
+
+    For Dhall __union types__ the correct value will be based on matching the type of JSON expression:
+
+> $ json-to-dhall 'List < Left : Text | Right : Integer >' <<< '[1, "bar"]'
+> [ < Right = +1 | Left : Text >, < Left = "bar" | Right : Integer > ]
```

Comment:
The translation should probably use the new union selector syntax because the union literal syntax will be deprecated in the next release of the standard:

```
$ json-to-dhall 'List < Left : Text | Right : Integer >' <<< '[1, "bar"]'
[ < Left : Text | Right : Integer >.Right +1, < Left : Text | Right : Integer >.Left "bar" > ]
```

Similar comment for the following union-related examples

Reply:
@Gabriel439 I'm a little confused here: Aren't the two different representations supposed to have the same type constructor  (as one would expect, at least after the normalization) and differ only at the pretty-printing stage? If not, isn't/shouldn't there be an isomorphic function translating between these two?

I notice that in shell:
```sh
$ dhall <<< '[ < Right = +1 | Left : Text >, < Left = "bar" | Right : Integer > ]'
[ < Right = +1 | Left : Text >, < Left = "bar" | Right : Integer > ]

$ dhall <<< '[ < Left : Text | Right : Integer >.Right +1, < Left : Text | Right : Integer >.Left "bar"
 ]'
[ < Left : Text | Right : Integer >.Right +1
, < Left : Text | Right : Integer >.Left "bar"
]
```

while in ghci (now I do see that they actually have different constructors):
```haskell

>>> import qualified Dhall.Parser as D
>>> import qualified Dhall.Core as D

>>> Right uNew <- traverse Dhall.Import.load $ D.normalize <$> D.exprFromText "" "[ < Left : Text | Right : Integer >.Right +1, < Left : Text | Right : Integer >.Left \"bar\" ]"                                                                                                                                       
>>> Right uOld <- traverse Dhall.Import.load $ D.normalize <$> D.exprFromText "" "[ < Right = +1 | Left : Text >, < Left = \"bar\" | Right : Integer > ]"

>>> uNew                                                                                                                                                  
ListLit Nothing (fromList [App (Field (Union (fromList [("Left",Just Text),("Right",Just Integer)])) "Right") (IntegerLit 1),App (Field (Union (fromList [("Left",Just Text),("Right",Just Integer)])) "Left") (TextLit (Chunks [] "bar"))])                                                                                  

>>> uOld                                                                                                                                                  
ListLit Nothing (fromList [UnionLit "Right" (IntegerLit 1) (fromList [("Left",Just Text)]),UnionLit "Left" (TextLit (Chunks [] "bar")) (fromList [("Right",Just Integer)])])
```

Do you think there should be a switch for opting out of the new notation provided by the command-line tool?
I guess that we will also have to factor out the Union type definition in lists of union values for the sake of readability, perhaps not this time around though. 
(BTW, for lists  `[ typ.v1, typ.v2,... ]` longer than one the type prefix is strictly redundant since it must be the same. On the other hand for empty lists it is necessary. Spontaneously, it would seem logical to somehow move the type outside of the list itself by default?)